# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/painting/nsDisplayList.cpp
# Commit: 33dbc637a003
# Full Hash: 33dbc637a003ff6431860440325a5ece32a1c91d
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2021-05-19 08:32:22
# Regressor Bug: 1540737
# File Overlap Count: 2
# Description:
#   Bug 1540737 - Allow filter painting to take a callback for painting children. r=miko
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D114670
# ==============================================================================

diff -r d7a58cff3e92 -r 33dbc637a003 layout/painting/nsDisplayList.cpp
--- a/layout/painting/nsDisplayList.cpp	Wed May 19 00:38:57 2021 +0000
+++ b/layout/painting/nsDisplayList.cpp	Wed May 19 01:11:30 2021 +0000
@@ -9906,12 +9906,37 @@
 
 void nsDisplayFilters::PaintAsLayer(nsDisplayListBuilder* aBuilder,
                                     gfxContext* aCtx, LayerManager* aManager) {
+  PaintWithContentsPaintCallback(aBuilder, aCtx, [&](gfxContext* aContext) {
+    BasicLayerManager* basic = aManager->AsBasicLayerManager();
+    RefPtr<gfxContext> oldCtx = basic->GetTarget();
+    basic->SetTarget(aContext);
+
+    aManager->EndTransaction(FrameLayerBuilder::DrawPaintedLayer, aBuilder);
+    basic->SetTarget(oldCtx);
+  });
+}
+
+void nsDisplayFilters::PaintWithContentsPaintCallback(
+    nsDisplayListBuilder* aBuilder, gfxContext* aCtx,
+    const std::function<void(gfxContext* aContext)>& aPaintChildren) {
   imgDrawingParams imgParams(aBuilder->GetImageDecodeFlags());
   nsRect borderArea = nsRect(ToReferenceFrame(), mFrame->GetSize());
   SVGIntegrationUtils::PaintFramesParams params(*aCtx, mFrame, GetPaintRect(),
-                                                borderArea, aBuilder, aManager,
+                                                borderArea, aBuilder, nullptr,
                                                 mHandleOpacity, imgParams);
-  SVGIntegrationUtils::PaintFilter(params);
+
+  gfxPoint userSpaceToFrameSpaceOffset =
+      SVGIntegrationUtils::GetOffsetToUserSpaceInDevPx(mFrame, params);
+
+  SVGIntegrationUtils::PaintFilter(
+      params,
+      [&](gfxContext& aContext, nsIFrame* aTarget, const gfxMatrix& aTransform,
+          const nsIntRect* aDirtyRect, image::imgDrawingParams& aImgParams) {
+        gfxContextMatrixAutoSaveRestore autoSR(&aContext);
+        aContext.SetMatrixDouble(aContext.CurrentMatrixDouble().PreTranslate(
+            -userSpaceToFrameSpaceOffset));
+        aPaintChildren(&aContext);
+      });
   nsDisplayFiltersGeometry::UpdateDrawResult(this, imgParams.result);
 }
 